README.rst
.. image:: https://dev.azure.com/spacetelescope/wiimatch/_apis/build/status/spacetelescope.wiimatch?branchName=master
  :target: https://dev.azure.com/spacetelescope/wiimatch/_build/latest?definitionId=13&branchName=master
  :alt: Azure Pipelines Status

.. image:: https://readthedocs.org/projects/wiimatch/badge/?version=latest
  :target: https://wiimatch.readthedocs.io/en/latest/?badge=latest
  :alt: Documentation Status

.. image:: https://codecov.io/gh/spacetelescope/wiimatch/branch/master/graph/badge.svg
  :target: https://codecov.io/gh/spacetelescope/wiimatch
  :alt: Codecov Status

``wiimatch`` is a package that provides core computational algorithms for
optimal "matching" of weighted N-dimensional image intensity data using
(multivariate) polynomials.


CHANGELOG.rst
.. _release_notes:

=============
Release Notes
=============

.. 0.3.3 (unreleased)
   ==================

0.3.2 (11-November-2023)
========================

- Maintenance release.


0.3.1 (20-July-2022)
====================

- Maintenance release.


0.3.0 (20-July-2022)
====================

- Added data containers module and updated main code to support these containers
  with the purpose of minimizing memory usage by writing/loading data
  arrays to temporary files when needed. [#21]


0.2.1 (08-July-2022)
====================

- Updated code to reduce warnings with latest ``numpy`` versions. [#16]

- Optimized code to improve performance and minimize memory usage when either
  ``masks`` and/or ``sigmas`` have default values. [#18, #19]


0.2.0 (07-August-2019)
======================

Added
^^^^^

- Added a new, more stable, solver ``rlu_solve()``. ``match_lsq()`` now takes
  a new parameter ``solver`` which, by default, is set to ``'LU'`` - the
  new solver. [#1]

Fixed
^^^^^

- Updated package structure, setup, docs. [#1]


0.1.2 (12-June-2017)
====================

Added
^^^^^

- Several functions now return more values that can be used to analyse returned
  results:

  - :py:func:`wiimatch.utils.create_coordinate_arrays` now returns effective
    ``center`` values used in generating coordinate array and coordinate system
    type (``'image'`` or ``'world'``);

  - :py:func:`wiimatch.lsq_optimizer.build_lsq_eqs` now returns coordinate
    arrays, effective ``center`` values used in generating coordinate array,
    and the coordinate system type of coordinates in addition to coefficients
    of linear equations;

  - :py:func:`wiimatch.match.match_lsq` now optionally returns coefficients
    of linear equations, coordinate arrays, effective ``center`` values used
    in generating coordinate array, and the coordinate system type of
    coordinates in addition to optimal solution to the matching problem.
    New parameter ``ext_return`` indicates to return extended information.


0.1.1 (06-June-2017)
====================

Added
^^^^^

- ``center_cs`` parameter to :py:func:`wiimatch.utils.create_coordinate_arrays`
  :py:func:`wiimatch.match.match_lsq` and
  :py:func:`wiimatch.lsq_optimizer.build_lsq_eqs` in order to allow
  specification of the coordinate system of the center
  (``'image'`` or ``'world'``) when it is explicitly set.

Fixed
^^^^^

- Broken logic in :py:func:`wiimatch.utils.create_coordinate_arrays` code
  for generating coordinate arrays.


0.1.0 (09-May-2017)
===================

Initial release.


wiimatch/tests/test_match.py
"""
A module containing unit tests for the ``wcsutil`` module.

:License: :doc:`../LICENSE`

"""
import pytest
import numpy as np

from wiimatch import match


@pytest.mark.parametrize('solver', ['RLU', 'PINV'])
def test_match_lsq_solver(solver):
    # simulate background image data:
    c = [1.32, 0.15, 0.62, 0, 0.74, 0, 0, 0]
    im1 = np.zeros((5, 5, 4), dtype=float)
    cbg = c[0] * np.ones_like(im1)  # constand background level image

    # add slope:
    ind = np.indices(im1.shape, dtype=float)
    im3 = cbg + c[1] * ind[0] + c[2] * ind[1] + c[4] * ind[2]

    mask = np.ones_like(im1, dtype=np.int8)
    sigma = np.ones_like(im1, dtype=float)

    p = match.match_lsq(
        [im1, im3], [mask, mask], [sigma, sigma],
        degree=(1, 1, 1), center=(0, 0, 0), solver=solver
    )

    assert np.allclose(-p[0], p[1], rtol=1.e-8, atol=1.e-12)
    assert np.allclose(c, 2 * np.abs(p[0]), rtol=1.e-8, atol=1.e-12)


def test_match_lsq_extended_return():
    # simulate background image data:
    c = [1.32, 0.15, 0.62, 0, 0.74, 0, 0, 0]
    im1 = np.zeros((5, 5, 4), dtype=float)
    cbg = c[0] * np.ones_like(im1)  # constand background level image

    # add slope:
    ind = np.indices(im1.shape, dtype=float)
    im3 = cbg + c[1] * ind[0] + c[2] * ind[1] + c[4] * ind[2]

    mask = np.ones_like(im1, dtype=np.int8)
    sigma = np.ones_like(im1, dtype=float)

    p, a, b, coord_arrays, eff_center, coord_system = match.match_lsq(
        [im1, im3], [mask, mask], [sigma, sigma],
        degree=1, center=(0, 0, 0), ext_return=True
    )

    assert np.allclose(-p[0], p[1], rtol=1.e-8, atol=1.e-12)
    assert np.allclose(c, 2 * np.abs(p[0]), rtol=1.e-8, atol=1.e-12)


@pytest.mark.parametrize('degree', [1, (1, 1, 1)])
def test_match_lsq_num_degree(degree):
    # simulate background image data:
    c = [1.32, 0.15, 0.62, 0, 0.74, 0, 0, 0]
    im1 = np.zeros((5, 5, 4), dtype=float)
    cbg = c[0] * np.ones_like(im1)  # constand background level image

    # add slope:
    ind = np.indices(im1.shape, dtype=float)
    im3 = cbg + c[1] * ind[0] + c[2] * ind[1] + c[4] * ind[2]

    mask = np.ones_like(im1, dtype=np.int8)
    sigma = np.ones_like(im1, dtype=float)

    p = match.match_lsq(
        [im1, im3], [mask, mask], [sigma, sigma],
        degree=degree, center=(0, 0, 0), solver='RLU', ext_return=False
    )

    assert np.allclose(-p[0], p[1], rtol=1.e-8, atol=1.e-12)
    assert np.allclose(c, 2 * np.abs(p[0]), rtol=1.e-8, atol=1.e-12)


wiimatch/tests/test_utils.py
"""
A module containing unit tests for the ``wcsutil`` module.

:License: :doc:`../LICENSE`

"""
import numpy as np

from wiimatch import utils


def test_utils_coordinates():
    image_shape = (3, 5, 4)
    center = (0, 0, 0)
    c = utils.create_coordinate_arrays(image_shape, center=center)
    ind = np.indices(image_shape, dtype=float)[::-1]

    assert np.allclose(c[0], ind, rtol=1.e-8, atol=1.e-12)
    assert np.allclose(c[1], center, rtol=1.e-8, atol=1.e-12)


def test_utils_coordinates_no_center():
    image_shape = (3, 5, 4)
    c = utils.create_coordinate_arrays(image_shape, center=None)
    ind = np.indices(image_shape, dtype=float)[::-1]

    center = tuple(i // 2 for i in image_shape)
    for orig, cc, i in zip(center, c[0], ind):
        assert np.allclose(i - orig, cc)

    assert np.allclose(c[1], center, rtol=1.e-8, atol=1.e-12)


wiimatch/utils.py
"""
This module provides utility functions for use by :py:mod:`wiimatch` module.

:Author: Mihai Cara (contact: help@stsci.edu)

:License: :doc:`LICENSE`

"""

import numpy as np

from .containers import WMInMemoryData


__all__ = ['create_coordinate_arrays']


def create_coordinate_arrays(image_shape, center=None, image2world=None,
                             center_cs='image',
                             container_cls=WMInMemoryData):
    """
    Create a list of coordinate arrays/grids for each dimension in the image
    shape. This function is similar to `numpy.indices` except it returns the
    list of arrays in reversed order. In addition, it can center image
    coordinates to a provided ``center`` and also convert image coordinates to
    world coordinates using provided ``image2world`` function.

    Parameters
    ----------
    image_shape : sequence of int
        The shape of the image/grid.

    center : iterable, None, optional
        An iterable of length equal to the number of dimensions in
        ``image_shape`` that indicates the center of the coordinate system
        in **image** coordinates when ``center_cs`` is ``'image'`` otherwise
        center is assumed to be in **world** coordinates (when ``center_cs``
        is ``'world'``). When ``center`` is `None` then ``center`` is
        set to the middle of the "image" as ``center[i]=image_shape[i]//2``.
        If ``image2world`` is not `None` and ``center_cs`` is ``'image'``,
        then supplied center will be converted to world coordinates.

    image2world : function, None, optional
        Image-to-world coordinates transformation function. This function
        must be of the form ``f(x,y,z,...)`` and accept a number of arguments
        `numpy.ndarray` arguments equal to the dimensionality of images.

    center_cs : {'image', 'world'}, optional
        Indicates whether ``center`` is in image coordinates or in world
        coordinates. This parameter is ignored when ``center`` is set to
        `None`: it is assumed to be `False`. ``center_cs`` *cannot be*
        ``'world'`` when ``image2world`` is `None` unless ``center`` is `None`.

    Returns
    -------
    coord_arrays : list
        A list of `numpy.ndarray` coordinate arrays each of ``image_shape``
        shape.

    eff_center : tuple
        A tuple of coordinates of the effective center as used in generating
        coordinate arrays.

    coord_system : {'image', 'world'}
        Coordinate system of the coordinate arrays and returned ``center``
        value.

    Examples
    --------
    >>> import wiimatch
    >>> wiimatch.utils.create_coordinate_arrays((3, 5, 4))   # doctest: +FLOAT_CMP
    ((array([[[-1.,  0.,  1.,  2.],
            [-1.,  0.,  1.,  2.],
            [-1.,  0.,  1.,  2.],
            [-1.,  0.,  1.,  2.],
            [-1.,  0.,  1.,  2.]],
           [[-1.,  0.,  1.,  2.],
            [-1.,  0.,  1.,  2.],
            [-1.,  0.,  1.,  2.],
            [-1.,  0.,  1.,  2.],
            [-1.,  0.,  1.,  2.]],
           [[-1.,  0.,  1.,  2.],
            [-1.,  0.,  1.,  2.],
            [-1.,  0.,  1.,  2.],
            [-1.,  0.,  1.,  2.],
            [-1.,  0.,  1.,  2.]]]),
    array([[[-2., -2., -2., -2.],
            [-1., -1., -1., -1.],
            [ 0.,  0.,  0.,  0.],
            [ 1.,  1.,  1.,  1.],
            [ 2.,  2.,  2.,  2.]],
           [[-2., -2., -2., -2.],
            [-1., -1., -1., -1.],
            [ 0.,  0.,  0.,  0.],
            [ 1.,  1.,  1.,  1.],
            [ 2.,  2.,  2.,  2.]],
           [[-2., -2., -2., -2.],
            [-1., -1., -1., -1.],
            [ 0.,  0.,  0.,  0.],
            [ 1.,  1.,  1.,  1.],
            [ 2.,  2.,  2.,  2.]]]),
    array([[[-2., -2., -2., -2.],
            [-2., -2., -2., -2.],
            [-2., -2., -2., -2.],
            [-2., -2., -2., -2.],
            [-2., -2., -2., -2.]],
           [[-1., -1., -1., -1.],
            [-1., -1., -1., -1.],
            [-1., -1., -1., -1.],
            [-1., -1., -1., -1.],
            [-1., -1., -1., -1.]],
           [[ 0.,  0.,  0.,  0.],
            [ 0.,  0.,  0.,  0.],
            [ 0.,  0.,  0.,  0.],
            [ 0.,  0.,  0.,  0.],
            [ 0.,  0.,  0.,  0.]]])), (1.0, 2.0, 2.0), 'image')

    """
    if center_cs not in ['image', 'world']:
        raise ValueError("Parameter 'center_cs' must take one of the "
                         "following two values: 'image' or 'world'.")

    if center is None:
        # set the center at the center of the image array:
        center = tuple([float(i // 2) for i in image_shape])
        center_cs = 'image'

    else:
        if len(center) != len(image_shape):
            raise ValueError("Number of coordinates of the 'center' must "
                             "match the dimensionality of the image.")

        if center_cs == 'world' and image2world is None:
            raise ValueError("'center_cs' cannot be 'world' when 'image2world'"
                             " is not defined.")

    ind = np.indices(image_shape, dtype=float)[::-1]

    if image2world is None:
        coord_system = 'image'
        eff_center = tuple([c for c in center])

    else:
        if center_cs == 'world':
            eff_center = tuple(map(float, center))

        else:
            # convert image's center from image to world coordinates:
            eff_center = tuple(map(float, image2world(*center)))

        # convert pixel indices to world coordinates:
        # TODO: get rid of the ravel/reshape dancing once
        #       issue https://github.com/spacetelescope/gwcs/issues/89
        #       is fixed.
        ind = image2world(*[x.ravel() for x in ind])
        ind = [x.reshape(image_shape) for x in ind]
        coord_system = 'world'

    coord_arrays = tuple([i - c for (i, c) in zip(ind, eff_center)])

    return coord_arrays, eff_center, coord_system


wiimatch/containers.py
"""
Data containers for accessing image data (i.e., `numpy.ndarray`)
uniformly whether they are kept in memory, as memory mapped files (load),
or stored to/loaded from a file as whole arrays.

:Author: Mihai Cara (contact: help@stsci.edu)

:License: :doc:`LICENSE`

"""

import abc
import tempfile

import numpy as np


__all__ = ['WMData', 'WMInMemoryData', 'WMMappedData', 'WMMemMappedData']


class WMData(abc.ABC):
    """ Base class for all data containers. Provides a common interface to
        access data.
    """
    kind = 'mapped'
    """ Hints to how data are stored: ``'mapped'``, ``'file'``, or
        ``'in-memory'``. May be used by code for performance optimization. """

    @property
    @abc.abstractmethod
    def data(self):
        """ Sets/Gets linked data.

        Parameters
        ----------
        data : object
            Data to be set.

        """
        pass

    @data.setter
    @abc.abstractmethod
    def data(self, data):
        pass

    @property
    @abc.abstractmethod
    def shape(self):
        """ Returns a tuple describing the shape of linked data. """
        pass


class WMInMemoryData(WMData):
    """ Acessor for in-memory `numpy.ndarray` data. """

    kind = 'in-memory'
    """ Hints to how data are stored: ``'mapped'``, ``'file'``, or
        ``'in-memory'``. May be used by code for performance optimization. """

    def __init__(self, data):
        super().__init__()
        self.data = data

    @property
    def data(self):
        """ Sets/gets linked `numpy.ndarray`.

        Parameters
        ----------
        data : object
            Data to be set.

        """
        return self._data

    @data.setter
    def data(self, data):
        self._data = np.asarray(data)

    @property
    def shape(self):
        """ Returns a tuple describing the shape of linked data. """
        return np.shape(self._data)


class WMMappedData(WMData):
    """ Data container for arrays stored in temporary files. This is best
        suited when array data are needed in memory all at once and when array
        is not needed - it can be stored to a file.

        To access small segments of data, use cls:`WMMemMappedData`.
    """

    kind = 'file'
    """ Hints to how data are stored: ``'mapped'``, ``'file'``, or
        ``'in-memory'``. May be used by code for performance optimization. """

    def __init__(self, data, tmpfile=None, prefix='tmp_wiimatch_',
                 suffix='.npy', tmpdir=''):
        super().__init__()
        if tmpfile is None:
            self._close = True
            self._tmp = tempfile.NamedTemporaryFile(
                prefix=prefix,
                suffix=suffix,
                dir=tmpdir
            )
            if not self._tmp:
                raise RuntimeError("Unable to create temporary file.")
        else:
            # temp file managed by the caller
            self._close = False
            self._tmp = tmpfile

        self.data = data

    @property
    def data(self):
        """ Sets/gets linked `numpy.ndarray`.

        Parameters
        ----------
        data : object
            Data to be set.

        """
        self._tmp.seek(0)
        return np.load(self._tmp)

    @data.setter
    def data(self, data):
        data = np.asarray(data)
        self._data_shape = data.shape
        self._tmp.seek(0)
        np.save(self._tmp, data)

    def __del__(self):
        if self._close:
            self._tmp.close()

    @property
    def shape(self):
        """ Returns a tuple describing the shape of linked data. """
        return self._data_shape


class WMMemMappedData(WMData):
    """ Data container for arrays stored in temporary files. This is best
        suited when array data are needed in memory all at once and when array
        is not needed - it can be stored to a file.

        To access entire data arrays, use cls:`WMMappedData`.
    """

    kind = 'mapped'
    """ Hints to how data are stored: ``'mapped'``, ``'file'``, or
        ``'in-memory'``. May be used by code for performance optimization. """

    def __init__(self, data, tmpfile=None, prefix='tmp_wiimatch_',
                 suffix='.npy', tmpdir=''):
        super().__init__()
        if tmpfile is None:
            self._close = True
            self._tmp = tempfile.NamedTemporaryFile(
                prefix=prefix,
                suffix=suffix,
                dir=tmpdir
            )
            if not self._tmp:
                raise RuntimeError("Unable to create temporary file.")
        else:
            # temp file managed by the caller
            self._close = False
            self._tmp = tmpfile

        self.data = data

    @property
    def data(self):
        """ Sets/gets linked `numpy.ndarray`.

        Parameters
        ----------
        data : object
            Data to be set.

        """
        return self._data

    @data.setter
    def data(self, data):
        data = np.asarray(data)
        self._data_shape = data.shape
        self._tmp.seek(0)
        self._data = np.memmap(self._tmp, dtype=data.dtype.type, mode='w+',
                               shape=data.shape)
        self._data[...] = data[...]

    def __del__(self):
        if self._close:
            self._data = None
            self._tmp.close()

    @property
    def shape(self):
        """ Returns a tuple describing the shape of linked data. """
        return self._data_shape


wiimatch/match.py
"""
A module that provides main API for optimal (LSQ) "matching" of weighted
N-dimensional image intensity data using (multivariate) polynomials.

:Author: Mihai Cara (contact: help@stsci.edu)

:License: :doc:`LICENSE`

"""
import numpy as np

from .lsq_optimizer import build_lsq_eqs, pinv_solve, rlu_solve
from .containers import WMData, WMInMemoryData, WMMappedData


__all__ = ['match_lsq']

SUPPORTED_SOLVERS = ['RLU', 'PINV']


def match_lsq(images, masks=None, sigmas=None, degree=0,
              center=None, image2world=None, center_cs='image',
              ext_return=False, solver='RLU',
              default_container=WMMappedData):
    r"""match_lsq(images, masks=None, sigmas=None, degree=0, center=None,
    image2world=None, center_cs='image', ext_return=False, solver='RLU',
    default_container=WMInMemoryData)

    Compute coefficients of (multivariate) polynomials that once subtracted
    from input images would provide image intensity matching in the least
    squares sense.

    Parameters
    ----------
    images : list of WMData and/or numpy.ndarray
        A list of `~wiimatch.containers.WMData` to 1D, 2D, etc.
        `numpy.ndarray` data arrays whose "intensities" must be "matched".
        All arrays must have identical
        shapes. When ``images`` is a list of `numpy.ndarray`, the container
        class specified by the ``default_container`` will be used to convert
        `numpy.ndarray` to `WMData` objects. Input list may mix `WMData`,
        `numpy.ndarray`, and `None` objects.

    masks : list of WMData and/or numpy.ndarray and/or None, None, optional
        A list of `WMData` of the same length as ``images``.
        Non-zero mask elements in data arrays indicate valid data in the
        corresponding ``images`` array. Mask arrays must have identical shape
        to that of the arrays in input ``images``. Default value of `None`
        indicates that all pixels in (the corresponding) input images are valid.
        When ``masks`` is a list of `numpy.ndarray`, the container class
        specified by the ``default_container`` will be used to convert
        `numpy.ndarray` to `WMData` objects. Input list may mix `WMData`,
        `numpy.ndarray`, and `None` objects.

    sigmas : list of WMData and/or numpy.ndarray and/or numbers, None, optional
        A list of `WMData` of the same length as ``images``
        representing the uncertainties of the data in the corresponding array
        in ``images``. Uncertainty arrays must have identical shape to that of
        the arrays in input ``images``. A numeric value for a ``sigmas``
        element will apply to all pixels in the corresponding ``images``
        element. The default value of `None` indicates
        that all pixels will be assigned equal weights. When ``sigmas``
        is a list of `numpy.ndarray`, the container class specified by the
        ``default_container`` will be used to convert `numpy.ndarray` to
        `WMData` objects. When ``sigmas`` is `None`, then all pixels
        in all images will be assigned weight 1.

    degree : iterable, int, optional
        A list of polynomial degrees for each dimension of data arrays in
        ``images``. The length of the input list must match the dimensionality
        of the input images. When a single integer number is provided, it is
        assumed that the polynomial degree in each dimension is equal to
        that integer.

    center : iterable, None, optional
        An iterable of length equal to the number of dimensions in
        ``image_shape`` that indicates the center of the coordinate system
        in **image** coordinates when ``center_cs`` is ``'image'`` otherwise
        center is assumed to be in **world** coordinates (when ``center_cs``
        is ``'world'``). When ``center`` is `None` then ``center`` is
        set to the middle of the "image" as ``center[i]=image_shape[i]//2``.
        If ``image2world`` is not `None` and ``center_cs`` is ``'image'``,
        then supplied center will be converted to world coordinates.

    image2world : function, None, optional
        Image-to-world coordinates transformation function. This function
        must be of the form ``f(x,y,z,...)`` and accept a number of arguments
        `numpy.ndarray` arguments equal to the dimensionality of images.

    center_cs : {'image', 'world'}, optional
        Indicates whether ``center`` is in image coordinates or in world
        coordinates. This parameter is ignored when ``center`` is set to
        `None`: it is assumed to be `False`. ``center_cs`` *cannot be*
        ``'world'`` when ``image2world`` is `None` unless ``center`` is `None`.

    ext_return : bool, optional
        Indicates whether this function should return additional values besides
        optimal polynomial coefficients (see ``bkg_poly_coeff`` return value
        below) that match image intensities in the LSQ sense. See **Returns**
        section for more details.

    solver : {'RLU', 'PINV'}, optional
        Specifies method for solving the system of equations.

    default_container : class
        A class that is a subclass of `WMData` that will be used to
        wrap input and internal `numpy.ndarray` arrays. Must be able to
        instantiate from a single argument - a data aray.

    Returns
    -------
    bkg_poly_coeff : numpy.ndarray
        When ``nimages`` is `None`, this function returns a 1D `numpy.ndarray`
        that holds the solution (polynomial coefficients) to the system.

        When ``nimages`` is **not** `None`, this function returns a 2D
        `numpy.ndarray` that holds the solution (polynomial coefficients)
        to the system. The solution is grouped by image.

    a : numpy.ndarray
        A 2D `numpy.ndarray` that holds the coefficients of the linear system
        of equations. This value is returned only when ``ext_return``
        is `True`.

    b : numpy.ndarray
        A 1D `numpy.ndarray` that holds the free terms of the linear system of
        equations. This value is returned only when ``ext_return`` is `True`.

    coord_arrays : list
        A list of `numpy.ndarray` coordinate arrays each of ``image_shape``
        shape. This value is returned only when ``ext_return`` is `True`.

    eff_center : tuple
        A tuple of coordinates of the effective center as used in generating
        coordinate arrays. This value is returned only when ``ext_return``
        is `True`.

    coord_system : {'image', 'world'}
        Coordinate system of the coordinate arrays and returned ``center``
        value. This value is returned only when ``ext_return`` is `True`.

    Notes
    -----
    :py:func:`match_lsq` builds a system of linear equations

    .. math::
        a \cdot c = b

    whose solution :math:`c` is a set of coefficients of (multivariate)
    polynomials that represent the "background" in each input image (these are
    polynomials that are "corrections" to intensities of input images) such
    that the following sum is minimized:

    .. math::
        L = \sum^N_{n,m=1,n \neq m} \sum_k
        \frac{\left[I_n(k) - I_m(k) - P_n(k) + P_m(k)\right]^2}
        {\sigma^2_n(k) + \sigma^2_m(k)}.

    In the above equation, index :math:`k=(k_1,k_2,...)` labels a position
    in input image's pixel grid [NOTE: all input images share a common
    pixel grid].

    "Background" polynomials :math:`P_n(k)` are defined through the
    corresponding coefficients as:

    .. math::
        P_n(k_1,k_2,...) = \sum_{d_1=0,d_2=0,...}^{D_1,D_2,...}
        c_{d_1,d_2,...}^n \cdot k_1^{d_1} \cdot k_2^{d_2}  \cdot \ldots .

    Coefficients :math:`c_{d_1,d_2,...}^n` are arranged in the vector :math:`c`
    in the following order:

    .. math::
        (c_{0,0,\ldots}^1,c_{1,0,\ldots}^1,\ldots,c_{0,0,\ldots}^2,
        c_{1,0,\ldots}^2,\ldots).

    :py:func:`match_lsq` returns coefficients of the polynomials that
    minimize *L*.

    Examples
    --------
    >>> import wiimatch
    >>> from wiimatch.containers import WMInMemoryData
    >>> import numpy as np
    >>> im1 = np.zeros((5, 5, 4), dtype=float)
    >>> cbg = 1.32 * np.ones_like(im1)
    >>> ind = np.indices(im1.shape, dtype=float)
    >>> im3 = cbg + 0.15 * ind[0] + 0.62 * ind[1] + 0.74 * ind[2]
    >>> mask = WMInMemoryData(np.ones_like(im1, dtype=np.int8))
    >>> sigma = WMInMemoryData(np.ones_like(im1, dtype=float))
    >>> wiimatch.match.match_lsq([WMInMemoryData(im1), WMInMemoryData(im3)],
    ... [mask, mask], [sigma, sigma], degree=(1,1,1), center=(0,0,0))  # doctest: +FLOAT_CMP
    array([[-6.60000000e-01, -7.50000000e-02, -3.10000000e-01,
            -6.96331881e-16, -3.70000000e-01, -1.02318154e-15,
            -5.96855898e-16,  2.98427949e-16],
           [ 6.60000000e-01,  7.50000000e-02,  3.10000000e-01,
             6.96331881e-16,  3.70000000e-01,  1.02318154e-15,
             5.96855898e-16, -2.98427949e-16]])

    """
    solver = solver.upper()
    if solver not in SUPPORTED_SOLVERS:
        ns = len(SUPPORTED_SOLVERS)
        raise ValueError("'solver' must be one of the supported solvers: '{}'"
                         .format(SUPPORTED_SOLVERS[0] if ns == 1 else
                                 '\', \''.join(SUPPORTED_SOLVERS[:-1]) +
                                 '\'' + (',' if ns > 2 else '') +
                                 ' or \'{}'.format(SUPPORTED_SOLVERS[-1])))

    images = [default_container(im) if isinstance(im, np.ndarray) else im
              for im in images]

    # check that all images have the same shape:
    shapes = set([])
    for im in images:
        shapes.add(im.shape)
    if len(shapes) > 1:
        raise ValueError("All images must have identical shapes.")

    nimages = len(images)
    ndim = len(images[0].shape)

    # check that the number of good pixel mask arrays matches the numbers
    # of input images, and if 'masks' is None - set all of them to True:
    if masks is not None:
        if len(masks) != nimages:
            raise ValueError("Length of masks list must match the length of "
                             "the image list.")

        for m in masks:
            if m is not None:
                shapes.add(m.shape)
        if len(shapes) > 1:
            raise ValueError("Shape of each mask array must match the shape "
                             "of input images.")

        # make a copy of the masks since we might modify these masks later
        masks = [default_container(m.copy()) if isinstance(m, np.ndarray) else m
                 for m in masks]

    # check that the number of sigma arrays matches the numbers
    # of input images, and if 'sigmas' is None - set all of them to 1:
    if sigmas is not None:
        nns = sum(int(s is None) for s in sigmas)
        if nns > 0:
            if nns != nimages:
                raise ValueError("'sigmas' must be either None, or a list of "
                                 "all None, or a list of all numpy.ndarray, "
                                 "WMData, and/or numbers")
            sigmas = None

        else:
            if len(sigmas) != nimages:
                raise ValueError("Length of sigmas list must match the length of "
                                 "the image list.")

            for s in sigmas:
                if np.ndim(s):
                    shapes.add(s.shape)
            if len(shapes) > 1:
                raise ValueError("Shape of each sigma array must match the shape "
                                 "of input images.")

            # make sure every element is a WMData:
            new_sigmas = []
            for s in sigmas:
                if isinstance(s, WMData):
                    new_sigmas.append(s)
                elif isinstance(s, np.ndarray) and np.ndim(s):
                    new_sigmas.append(default_container(s))
                else:
                    new_sigmas.append(WMInMemoryData(s))
            sigmas = new_sigmas

    # check that 'degree' has the same length as the number of dimensions
    # in image arrays:
    if hasattr(degree, '__iter__'):
        if len(degree) != ndim:
            raise ValueError("The length of 'degree' parameter must match "
                             "the number of image dimensions.")
        degree = tuple([int(d) for d in degree])

    else:
        intdeg = int(degree)
        degree = tuple([intdeg for i in range(ndim)])

    # check that 'center' has the same length as the number of dimensions
    # in image arrays:
    if hasattr(center, '__iter__'):
        if len(center) != ndim:
            raise ValueError("The length of 'center' parameter must match "
                             "the number of image dimensions.")

    elif center is not None:
        center = tuple([center for i in range(ndim)])

    # build the system of equations:
    a, b, coord_arrays, eff_center, coord_system = build_lsq_eqs(
        images,
        masks,
        sigmas,
        degree,
        center=center,
        image2world=image2world,
        center_cs=center_cs
    )

    # solve the system:
    if solver == 'RLU':
        bkg_poly_coef = rlu_solve(a, b, nimages)
    else:
        tol = np.finfo(images[0].data.dtype).eps**(2.0 / 3.0)
        bkg_poly_coef = pinv_solve(a, b, nimages, tol)

    if ext_return:
        return bkg_poly_coef, a, b, coord_arrays, eff_center, coord_system
    else:
        return bkg_poly_coef
